sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -130,6 +130,13 @@ class StreamExecution(
   protected var offsetSeqMetadata = OffsetSeqMetadata(
     batchWatermarkMs = 0, batchTimestampMs = 0, sparkSession.conf)
 
+  /**
+   * A map from watermarked attributes to their current watermark. The minimum watermark
+   * timestamp present here will be used as the overall query watermark in offsetSeqMetadata;
+   * the query watermark is what's logged and used to age out old state.
+   */
+  protected var attributeWatermarkMsMap: AttributeMap[Long] = AttributeMap(Seq())
```

Comment:
this is needed only inside a single function right? so make it a local variable. 
even better, if you dont have to make it a var, make it a val in the usual functional way.

Reply:
This map has to persist and get updated across batches, and I'm not sure how to do that with a local variable or a val.

Reply:
I see. I was mistaken. However in that case, use a mutable.HashMap instead of var. That the code style we use with scala is not to use vars unless absolutely needed.